File: sample/sample-android/src/main/java/com/example/sampleandroid/MainActivity.kt
//todo change package name
package com.example.sampleandroid

import android.os.Bundle
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.platform.setContent
import androidx.compose.ui.tooling.preview.Preview
import com.example.sampleandroid.ui.AndroidLibraryTheme
import com.example.sampleandroid.view.MainScaffold
import com.example.sampleandroid.view.home.HomeScreen
import com.example.sampleandroid.view.model.*
import kim.jeonghyeon.androidlibrary.compose.asValue
import kim.jeonghyeon.client.BaseActivity
import kim.jeonghyeon.client.BaseViewModel
import kim.jeonghyeon.client.Deeplink
import kim.jeonghyeon.sample.deeplinkList
import kim.jeonghyeon.sample.viewmodel.*

class MainActivity : BaseActivity() {
    override val rootViewModel: BaseViewModel = HomeViewModel()

    override val deeplinks: List<Deeplink> = deeplinkList

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setViewModelContent { viewModel ->
            AndroidLibraryTheme {
                MainScaffold {
                    Surface(color = MaterialTheme.colors.background) {
                        ScreenContent(viewModel)
                    }
                }
            }
        }
    }

    //todo remove to library if composable on library is available
    fun setViewModelContent(content: @Composable (model: BaseViewModel) -> Unit) {
        setContent {
            currentViewModel.asValue()?.let {
                @Suppress("EXPERIMENTAL_API_USAGE")
                it.onCompose()
                content(it)
            }
        }
    }
}


@Composable
fun ScreenContent(viewModel: BaseViewModel) = when (viewModel) {
    is HomeViewModel -> HomeScreen(viewModel)
    is ApiSingleViewModel -> ApiSingleScreen(viewModel)
    is ApiAnnotationViewModel -> ApiAnnotationScreen(viewModel)
    is ApiBindingViewModel -> ApiBindingScreen(viewModel)
    is ApiDbViewModel -> ApiDbScreen(viewModel)
    is ApiExternalViewModel -> ApiExternalScreen(viewModel)
    is ApiHeaderViewModel -> ApiHeaderScreen(viewModel)
    is ApiParallelViewModel -> ApiParallelScreen(viewModel)
    is ApiPollingViewModel -> ApiPollingScreen(viewModel)
    is ApiSequentialViewModel -> ApiSequentialScreen(viewModel)
    is DbSimpleViewModel -> DbSimpleScreen(viewModel)
    is DeeplinkViewModel -> DeeplinkScreen(viewModel)
    is DeeplinkSubViewModel -> DeeplinkSubScreen(viewModel)
    is SearchViewModel -> SearchScreen(viewModel)
    is SignInViewModel -> SignInScreen(viewModel)
    is SignUpViewModel -> SignUpScreen(viewModel)
    is UserViewModel -> UserScreen(viewModel)
    is RetrofitViewModel -> RetrofitScreen(viewModel)
    else -> error("Screen is not defined for $viewModel ")
}

@Preview(showBackground = true)
@Composable
fun DefaultPreview() {
    HomeScreen(HomeViewModel())
}

File: sample/sample-native/kotlinIOS/screen/MainActivity.swift
//
//  MainActivity.swift
//  kotlinIOS
//
//  Created by hyun kim on 25/10/20.
//  Copyright © 2020 hyun kim. All rights reserved.
//

import Foundation
import sample_base
import SwiftUI

func MainActivity() -> some View {
    BaseActivity(rootViewModel: HomeViewModel(), deeplinks: DeeplinkKt.deeplinkList) { viewModel in
        //you can customize to add tab view or drawer. etc..
        Box {
            switch (viewModel) {
            case is HomeViewModel: HomeScreen(viewModel as! HomeViewModel)
            case is ApiSingleViewModel: ApiSingleScreen(viewModel as! ApiSingleViewModel)
            case is ApiAnnotationViewModel: ApiAnnotationScreen(viewModel as! ApiAnnotationViewModel)
            case is ApiBindingViewModel: ApiBindingScreen(viewModel as! ApiBindingViewModel)
            case is ApiDbViewModel: ApiDbScreen(viewModel as! ApiDbViewModel)
            case is ApiExternalViewModel: ApiExternalScreen(viewModel as! ApiExternalViewModel)
            case is ApiHeaderViewModel: ApiHeaderScreen(viewModel as! ApiHeaderViewModel)
            case is ApiParallelViewModel: ApiParallelScreen(viewModel as! ApiParallelViewModel)
            case is ApiPollingViewModel: ApiPollingScreen(viewModel as! ApiPollingViewModel)
            case is ApiSequentialViewModel: ApiSequentialScreen(viewModel as! ApiSequentialViewModel)
            case is DbSimpleViewModel: DbSimpleScreen(viewModel as! DbSimpleViewModel)
            case is DeeplinkViewModel: DeeplinkScreen(viewModel as! DeeplinkViewModel)
            case is DeeplinkSubViewModel: DeeplinkSubScreen(viewModel as! DeeplinkSubViewModel)
            case is SearchViewModel: SearchScreen(viewModel as! SearchViewModel)
            case is SignInViewModel: SignInScreen(viewModel as! SignInViewModel)
            case is SignUpViewModel: SignUpScreen(viewModel as! SignUpViewModel)
            case is UserViewModel: UserScreen(viewModel as! UserViewModel)
            default: EmptyView()
            }
        }
    }
}
